V2 docs (#93)
cohere-ai/cohere-developer-experience: fern/pages/v2/text-generation/migrating-v1-to-v2.mdx

]
```

### v1

```python PYTHON
message = "Are there fitness-related benefits?"

res_v1 = co_v1.chat(model="command-r-plus-08-2024",
                    message=message,
                    documents=documents)

print(res_v1.text)
```
```
Yes, we offer gym memberships and on-site yoga classes as part of our health and wellness benefits.
```

### v2

```python PYTHON
message = "Are there fitness-related benefits?"

messages = [{'role': 'user', 'content': message, "documents": documents}]
    
res_v2 = co_v2.chat(model="command-r-plus-08-2024",
                    messages=messages)

print(res_v2.message.content[0].text)
```
```
Yes, we offer gym memberships and on-site yoga classes as part of our health and wellness benefits.
```

## Search query generation

- Search query generation:
   - v1: Uses `search_queries_only` parameter
   - v2: No direct equivalent, alternative approaches available

### v1

```python PYTHON
res = co_v1.chat(model="command-r-plus-08-2024",
              message="Google and Apple revenue 2023",
              search_queries_only=True)
```

> ### v2

Author: done

image: "../../../assets/images/b3c8253-cohere_meta_image.jpg"  
keywords: "Cohere, text generation, LLMs, generative AI"

createdAt: "Thu Feb 29 2024 18:13:25 GMT+0000 (Coordinated Universal Time)"
updatedAt: "Thu May 23 2024 04:32:10 GMT+0000 (Coordinated Universal Time)"
---
This guide serves as a reference for developers looking to update their existing Cohere API v1 implementations to the new v2 standard.

It outlines the key differences and necessary changes when migrating from Cohere API v1 to v2.

It covers various aspects of the API, including chat functionality, RAG (Retrieval-Augmented Generation), and tool use.

Each section provides code examples for both v1 and v2, highlighting the structural changes in request formats, response handling, and new features introduced in v2.

```python PYTHON
# ! pip install -U cohere

import json
import cohere

co_v1 = cohere.Client(api_key="<YOUR API KEY>")
co_v2 = cohere.ClientV2(api_key="<YOUR API KEY>")
```

# Chat

## Messages

- Message structure: 
   - v1: uses separate `preamble` and `message` parameters.
   - v2: uses a `messages` list with role-based dictionaries (`system`, `user`, `assistant`, or `tool`).

> - Chat history:

Reviewer: i think `chat_history` can be included in the section above, since it's talking about the `messages` parameter

   - v1: manages chat history internally and uses a `chat_history` parameter.
   - v2: manual management of the conversation in the `messages` list.

### v1

> 

Reviewer: ```suggestion
```

```python PYTHON

> preamble = """## Task and Context
> You respond in concise sentences."""

Reviewer: ```suggestion
preamble = "You respond in concise sentences."
```

message_turn1 = "I'm joining a new startup called Co1t today. Could you help me write a one-sentence introduction message to my teammates."

res_turn1 = co_v1.chat(model="command-r-plus",
               preamble=preamble,
               message=message_turn1)

message_turn2 = "Make it shorter"

res_turn2 = co_v1.chat(model="command-r-plus",
                preamble=preamble,
                message=message_turn2,
                chat_history=res_turn1.chat_history)

for item in res_turn2.chat_history:
    print(item, "\n")
```

```
message="I'm joining a new startup called Co1t today. Could you help me write a one-sentence introduction message to my teammates." tool_calls=None role='USER' 

message='"Hi, I\'m [Your Name] and I\'m thrilled to join the Co1t team today as a [Your Role], eager to contribute my skills and ideas to the company\'s growth and success."' tool_calls=None role='CHATBOT' 

message='Make it shorter' tool_calls=None role='USER' 

message='"Excited to join Co1t today as a [Your Role] and looking forward to collaborating with the team."' tool_calls=None role='CHATBOT' 
```

> 

Reviewer: lots of spaces

messages = [
    {"role": "system", "content": instructions},
    {"role": "user", "content": query},
]

model = "command-r-plus-08-2024"

# Generate search queries (if any)
response = co_v2.chat(model=model, messages=messages, tools=web_search_tool)

search_queries = []

while response.message.tool_calls:

    print("Tool plan:")
    print(response.message.tool_plan, "\n")
    print("Tool calls:")
    for tc in response.message.tool_calls:
        print(f"Tool name: {tc.function.name} | Parameters: {tc.function.arguments}")
    print("=" * 50)

    messages.append(
        {
            "role": "assistant",
            "tool_calls": response.message.tool_calls,
            "tool_plan": response.message.tool_plan,
        }
    )

    # Step 3: Get tool results
    tool_content = []
    for idx, tc in enumerate(response.message.tool_calls):
        tool_result = web_search(**json.loads(tc.function.arguments))
        tool_content.append(json.dumps(tool_result))
        messages.append(

>             {"role": "tool", "tool_call_id": tc.id, "tool_content": tool_content}

Reviewer: ```suggestion
            {"role": "tool", "tool_call_id": tc.id, "content": tool_content}
```

                    tools=tools_v1)

print(res_v1.tool_calls)
```
```
[ToolCall(name='get_weather', parameters={'location': 'Toronto'})]
```

**v2**

```python PYTHON
messages = [{"role": "user", "content": "What's the weather in Toronto?"}]

res_v2 = co_v2.chat(model="command-r-plus-08-2024", messages=messages, tools=tools_v2)

if res_v2.message.tool_calls:
    messages.append(
        {
            "role": "assistant",
            "tool_calls": res_v2.message.tool_calls,
            "tool_plan": res_v2.message.tool_plan,
        }
    )

    print(res_v2.message.tool_calls)
```
```
[ToolCallV2(id='get_weather_k88p0m8504w5', type='function', function=ToolCallV2Function(name='get_weather', arguments='{"location":"Toronto"}'))]
```

## Tool call ID

- v1: Tool calls do not emit tool call IDs
- v2: Tool calls emit tool call IDs. This will help the model match tool results to the right tool call.

**v1**
```
"tool_results": [
  {
    "call": {
      "name": <tool name>,
      "parameters": {
        <param name>: <param value>
      }
    },
    "outputs": [{
      <key>: <value>
    }]
  },
]

```
**v2**
```

"messages": [
  {
    "role": "tool", 
    "tool_call_id": "123", 

>     "tool_content": [

Reviewer: ```suggestion
    "content": [
```

  }
]
```

## Tool execution

- Function name access
   - v1: Access through `tc.name`
   - v2: Access through `tc.function.name`
   
- Function parameter access
   - v1: Access through `tc.parameters`
   - v2: Access through `tc.function.arguments`

- Chat history management
   - v1: Append `call` and `outputs` to the chat history
   - v2: Append `tool_call_id` and `tool_content` to `messages` to the chat history

**v1**

```python PYTHON
tool_content_v1 = []
if res_v1.tool_calls:
    for tc in res_v1.tool_calls:
        tool_call = {"name": tc.name, "parameters": tc.parameters}
        tool_result = functions_map[tc.name](**tc.parameters)
        tool_content_v1.append({"call": tool_call, "outputs": [tool_result]})
        
print(tool_content_v1)
```
```
[{'call': {'name': 'get_weather', 'parameters': {'location': 'Toronto'}}, 'outputs': [{'temperature': '20C'}]}]
```

**v2**

```python PYTHON
tool_content_v2 = []
if res_v2.message.tool_calls:
    for tc in res_v2.message.tool_calls:
        tool_result = functions_map[tc.function.name](
            **json.loads(tc.function.arguments)
        )
        tool_content_v2.append(json.dumps(tool_result))
        messages.append(

>             {"role": "tool", "tool_call_id": tc.id, "tool_content": tool_content_v2}

Reviewer: ```suggestion
            {"role": "tool", "tool_call_id": tc.id, "content": tool_content_v2}
```

        )

print(tool_content_v2)
```
```
['{"temperature": "20C"}']
```

> ## Response generation

Reviewer: I found this section confusing. i would combine it with the previous section

createdAt: "Thu Feb 29 2024 18:13:25 GMT+0000 (Coordinated Universal Time)"
updatedAt: "Thu May 23 2024 04:32:10 GMT+0000 (Coordinated Universal Time)"
---
This guide serves as a reference for developers looking to update their existing Cohere API v1 implementations to the new v2 standard.

It outlines the key differences and necessary changes when migrating from Cohere API v1 to v2.

It covers various aspects of the API, including chat functionality, RAG (Retrieval-Augmented Generation), and tool use.

Each section provides code examples for both v1 and v2, highlighting the structural changes in request formats, response handling, and new features introduced in v2.

```python PYTHON
# ! pip install -U cohere

import json
import cohere

co_v1 = cohere.Client(api_key="<YOUR API KEY>")
co_v2 = cohere.ClientV2(api_key="<YOUR API KEY>")
```

# Chat

## Messages

- Message structure: 
   - v1: uses separate `preamble` and `message` parameters.
   - v2: uses a single `messages` parameter consisting of a list of roles (`system`, `user`, `assistant`, or `tool`). The `system` role in v2 replaces the `preamble` parameter in v1.

- Chat history:

>    - v1: manage the chat history via the `chat_history` parameter.
>    - v2: manage the chat history via the `messages` list.

Reviewer: ```suggestion
   - v1: manages the chat history via the `chat_history` parameter.
   - v2: manages the chat history via the `messages` list.
```

**v1**

```python PYTHON
preamble = "You respond in concise sentences."

> 
> message_turn1 = "I'm joining a new startup called Co1t today. Could you help me write a one-sentence introduction message to my teammates."
> 
> res_turn1 = co_v1.chat(model="command-r-plus-08-2024",
>                preamble=preamble,
>                message=message_turn1)
> 
> message_turn2 = "Make it shorter"
> 
> res_turn2 = co_v1.chat(model="command-r-plus-08-2024",
>                 preamble=preamble,
>                 message=message_turn2,
>                 chat_history=res_turn1.chat_history)

Reviewer: this code snippet is hard to read.

```suggestion
res = co_v1.chat(
    model="command-r-plus-08-2024",
    preamble="You respond in concise sentences.",
    chat_history=[
        {
            "role": "user",
            "message": "Hello"
        },
        {
            "role": "chatbot",
            "message": "Hi, how can I help you today?"
        }
    ],
    message="What is the best way to introduce myself to my new teammates?")
```

message='Make it shorter' tool_calls=None role='USER' 

message='"Excited to join Co1t today as a [Your Role] and looking forward to collaborating with the team."' tool_calls=None role='CHATBOT' 
```

**v2**

```python PYTHON
system_message = "You respond in concise sentences."

message_turn1 = "I'm joining a new startup called Co1t today. Could you help me write a one-sentence introduction message to my teammates."

messages = [{'role': 'system', 'content': system_message},
           {'role': 'user', 'content': message_turn1}]

res_turn1 = co_v2.chat(model="command-r-plus-08-2024",
                    messages=messages)

message_turn2 = "Make it shorter"

messages.extend([{'role': 'assistant', 'content': res_turn1.message.content[0].text},
                 {'role': 'user', 'content': message_turn2}])

res_turn2 = co_v2.chat(model="command-r-plus-08-2024",
                          messages=messages)

messages.append({'role': 'assistant', 'content': res_turn2.message.content[0].text})

for message in messages:

>     print(message, "\n")

Reviewer: Same here, we should simplify:

```
res = co_v2.chat(
    model="command-r-plus-08-2024",
    messages=[
        {
            "role": "system",
            "content": "You respond in concise sentences."
        },
        {
            "role": "user",
            "content": "Hello"
        },
        {
            "role": "assistant",
            "content": "Hi, how can I help you today?"
        },
        {
            "role": "user",
            "content": "What is the best way to introduce myself to my new teammates?"
        }
    ])
```

res_v2 = co_v2.chat(model="command-r-plus-08-2024",
                    messages=messages,
                    tools=tools_v2)

if res_v2.message.tool_calls:
    messages.append({'role': 'assistant', 'tool_calls': res_v2.message.tool_calls, 'tool_plan': res_v2.message.tool_plan})
    
    print(res_v2.message.tool_calls)
```
```
[ToolCall2(id='get_weather_s8qdenha4v1z', type='function', function=ToolCall2Function(name='get_weather', arguments='{"location":"Toronto"}'))]
```

## Tool call ID

- v1: Tool calls do not emit tool call IDs
- v2: Tool calls emit tool call IDs. This will help the model match tool results to the right tool call.

**v1**
```
"tool_results" = [
  {
    "call": {
      "name": <tool name>,
      "parameters": {
        <param name>: <param value>
      }
    },
    "outputs": [{
      <key>: <value>
    }]
  },
]

```
**v2**
```
"messages": [
    {
    "role": "tool", 
    "tool_call_id": "123", 
    "tool_content": [
        # Option 1: Tool results as one string for one citation output
        {"type": "text", "text": "[{<key>: <value>}]"},
        # Option 2: Tool results with metadata for multiple citations 

>         {"type": "document", "document": {"id": "123", "content": "[{<key>: <value>}]", "metadata": {...}}},

Author: fixed
